### PR TITLE
Refactor worker 

### PR DESCRIPTION
**Pull request by ctring**
There are a lot of changes in this pr and it is quite hard to compare between the old and new code so I suggest just read the new code. Other than adding the ABORT command for explicitly aborting a transaction, this pr refactors the worker to make clearer the different phases that a transaction will go through. The code for each phase is put into its own well-defined method to make it easier to read and extend. The phases are:
```
READ_LOCAL_STORAGE,
WAIT_REMOTE_READ,
EXECUTE,
COMMIT,
FINISH,
```

`Worker::ProcessWorkerRequest` now only initializes the state of a new transaction.
`Worker::ProcessRemoteReadResult` is the same as before.

Both of these methods do not call other transaction processing methods but only create/change the state of a transaction. 

`Worker::AdvanceTransaction` is a new method, which calls the appropriate transaction processing method based on a transaction's state. This method is called whenever a new message arrives at the worker.

### CODE REVIEW

**Comment by johannkm** [comment]
I would benefit from comments explaining what the read and write sets contain at the point they're used, since they go from all keys to only keys in partition then back to all keys

**Comment by johannkm** [comment]
In commands.cpp, could the second if be replaced by a check?
```
if (cmd_ == "SET") {
      if (write_set.contains(args_[0])) {
```

File: module/scheduler_components/worker.cpp [hunk 1]
```diff
@@ -49,34 +55,143 @@ void Worker::Loop() {
...
+    }
+  }
+
+  // Create a state for the new transaction
+  auto state = txn_states_.emplace(
+      std::piecewise_construct,
+      std::forward_as_tuple(txn_id),
+      std::forward_as_tuple(txn_holder));
+
+  CHECK(state.second) << "Transaction " << txn_id << " has already been dispatched to this worker";
```
Comment (johannkm) [inline module/scheduler_components/worker.cpp:119]: This checks if Transaction state has been initialized?

Comment (ctring) [reply module/scheduler_components/worker.cpp:119]: Yes. This is just another cautionary check. Later I'll replace all such checks with `DCHECK`, which would be optimized out when we compile it in release mode.

File: module/scheduler_components/commands.cpp [hunk 2]
```diff
@@ -40,6 +40,10 @@ void KeyValueCommands::Execute(Transaction& txn) {
       if (read_set.contains(src) && write_set.contains(dst)) {
         write_set[dst] = read_set.at(src);
       }
+    } else if (cmd_ == "ABORT") {
+      if (write_set.contains(args_[0]) || read_set.contains(args_[0])) {
```
[the remark this answers is not in the document]
Comment (ctring) [reply module/scheduler_components/commands.cpp:44]: Nope. Previously, I added the key so that we could "control" which partition to abort but then I realized all partitions ran the same code. Actually this "ABORT" command is not needed either. I can just simply have an `else` here and abort for any token that we don't understand.

**Comment by ctring** [comment]
The way I look at this is that we have a pipeline that we want to push the transaction through. The `FALLTHROUGH` is a way to say that we want to push this transaction through as much as possible. However, there may be points on the pipeline where the transaction cannot push further until some external events happen. A messages coming into the worker are the driving force that push the transaction through this pipeline. If a txn doesn't need remote reads, the first message can push it completely through this pipeline. On the other hand, it has to be stuck at remote read wait, and each remote read message will chip away the barrier between that phase and the next. 

The `switch ... case ... fallthrough` idiom has two advantages:
  + You can control which phase on the pipeline that you want to jump to in a more "standardized" fashion; whereas calling the function directly would require ad-hoc if-then at the place you want to jump
  + When a transaction goes to sleep to wait for something, you can wake it up at the point where it left off simply by assigning the next phase.

I actually decided to refactor this because adding the confirmation wait code would result in two places where the txn would go to sleep then wake up, making the code flow hard to follow if they were implemented in an ad-hoc way. Removing the confirmation wait took the wind out of this PR, but I think this is still useful for later extension.

File: module/scheduler_components/worker.cpp (hunk 1 above)
Comment (johannkm) [reply module/scheduler_components/worker.cpp:119]: Gotcha, didn't realize CHECK wasn't.

File: module/scheduler_components/commands.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (johannkm) [reply module/scheduler_components/commands.cpp:44]: ABORT <key> <value> would kind of make sense to me. We could abort if the key != value, so we'd have primitive logical aborts. On the other hand it doesn't affect performance, so maybe we should just keep things simple.

**Comment by ctring** [comment]
> 
> 
> In commands.cpp, could the second if be replaced by a check?
> 
> ```
> if (cmd_ == "SET") {
>       if (write_set.contains(args_[0])) {
> ```

Replace them to CHECK is only safe for read set. When a partition that didn't possess a certain key, it would not have that key in the write set at this point. The command that would modify that key must be ignored, otherwise it would create a new key, which would later be committed to the wrong partition. In other words, it is still valid to not have a key in the write set while the command acting on that key is ignored.
